### PR TITLE
Story #12622: Developer doc & tools for CAS delegated authentication

### PR DESCRIPTION
**Pull request by bbenaissa**
## Description

L'objectif de cette PR est de fournir une documentation de developpeurs et outils pour tester la délégation d'authetification en SAML et OIDC. 

## Contributeur

VAS (Vitam Accessible en Service)

### CODE REVIEW

File: cas/cas-server/README.md [hunk 1]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
 - Only ONE organization may use x509 certificate authentication per email domain; CAS has no means to know to which
   organization a certificate belongs.
 
-## OpenId Connect authentication delegation
+## OpenId Connect (OIDC) authentication delegation
 
-> TODO
+VitamUI offers the possibility to delegate authentication to an external SSO using the OpenID Connect protocol.
```
Comment (ghost) [inline cas/cas-server/README.md]: Idp

File: cas/cas-server/README.md [hunk 2]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+## OpenId Connect (OIDC) authentication delegation
 
-> TODO
+VitamUI offers the possibility to delegate authentication to an external SSO using the OpenID Connect protocol. 
+To configure an external provider in VitamUI, you will need to follow the procedure below:
 
-## SAML authentication delegation
+- Create an OIDC provider for the domain mail.
+- As we are using the email address as users login, you put as email attribute the value `email`
+- Choosing the default email domain concerned by the provider.
```
Comment (ghost) [inline cas/cas-server/README.md]: ?

File: cas/cas-server/README.md [hunk 3]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+  - `Valid redirect URIs`
+  - `Valid post logout redirect URIs`
+  - `Web origins`
+  
+- The external provider must be accessible from CAS VM, at least the `Discovery URL`.
+- You need to restart the CAS service after adding the provider.
+
+**Settings example for local test**
+
+Bellow an example on OIDC authentication provider based on a all ready keycloak.
```
Comment (ghost) [inline cas/cas-server/README.md]: all ready --> pre-configured

File: cas/cas-server/README.md [hunk 4]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+
+To test the authentication delegation in the OIDC protocol, you will find an example of ready-made configuration here.
+- Simply launch the `run-dev.sh` script,
```
Comment (ghost) [inline cas/cas-server/README.md]: path please

File: cas/cas-server/README.md [hunk 5]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+    - secret: `QQXbm6947N5kYVL0yLDAHwlo3ZW2I8ui`
+    - OpenID Endpoint Configuration: `http://localhost:8041/realms/vitamui-test/.well-known/openid-configuration`
+- The user created has the following informations:
+    - Username: `demo@change-me.fr`
+    - Email: `demo@change-me.fr`
+    - Name: `demo oidc vitamui`
+    - Password: `ChangeIt.2024`
+
+** Vitamui provider collection content **
+Based on the test container, bellow an example on OIDC authentication provider for the local provider:
```
Comment (ghost) [inline cas/cas-server/README.md]: of

File: cas/cas-server/README.md [hunk 6]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+    customerId: '5c7927af7884583d1ebb6e7a74547a15e35d431599d976a9708eb12d6c5e56c9',
+    _class: 'providers'
+}
+
+```
+
+
+## SAML V2 authentication delegation
+
+To setting up Saml V2 authentication with vitamui,we follow the following steps:
```
Comment (ghost) [inline cas/cas-server/README.md]: set*

, please

File: cas/cas-server/README.md [hunk 7]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+
+- Retrieve the federation certificate from your IAM (e.g., orga-SAML.crt).
```
Comment (ghost) [inline cas/cas-server/README.md]: Idp

File: cas/cas-server/README.md [hunk 8]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+              ```yaml
+            cas.server.name: http://localhost:8383
+            cas.server.prefix: http://localhost:8383/cas
+            server.port: 8383
+            server.ssl.enabled: false
+            cas.server.tomcat.httpProxy.enabled: false
+            cas.server.tomcat.http:
+              - enabled: false
+             ```
+          - add a user sample and his password:
```
Comment (ghost) [inline cas/cas-server/README.md]: test user

File: cas/cas-server/README.md [hunk 9]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+            ```yaml
+              cas.authn.accept.users: myusernam@mydomainmail.fr::mypassword
+              cas.authn.attribute-repository.stub.attributes.email: myusernam@mydomainmail.fr
+             ```
+          - **myusernam@mydomainmail.fr**  is the user email for testing, and **mypassword** is the password
+
+      - run ```mvn clean package``` on the project.
+      - run ```java -jar target/cas.war``` 
+      - After launching,cas will generate some settings files inside the directory ```/etc/cas/saml```
```
Comment (ghost) [inline cas/cas-server/README.md:346]: ,

File: cas/cas-server/README.md [hunk 10]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+      - Goto Vitamui, create a Saml provider with the following informations:
```
Comment (ghost) [inline cas/cas-server/README.md]: Log in as a superadmin in VitamUI, and create...

File: cas/cas-server/README.md [hunk 11]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+        - Pattern: email domain configured before: ```mydomainmail.fr```
+        - Type: ```SAML```
+        - Email attribute: empty (except if the cas send the email after authentication)
+        - CAS Keystore: for testing: you upload any keystore with the right passowrd.
+        - IDP Metadata: upload the file generated by running external CAS, from the path ```/etc/cas/saml/idp-metadata.xml```
+        - Assertions : false
+        - Signed request: false.
+      - On vitamui, after creating the provider (SSO list) : 
+        - we download the ```SPS-metadata.xml``` file and copy it in the vm of external CAS, in any directory,
```
Comment (ghost) [inline cas/cas-server/README.md]: remove vm*

File: cas/cas-server/README.md [hunk 12]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+        - Pattern: email domain configured before: ```mydomainmail.fr```
+        - Type: ```SAML```
+        - Email attribute: empty (except if the cas send the email after authentication)
+        - CAS Keystore: for testing: you upload any keystore with the right passowrd.
+        - IDP Metadata: upload the file generated by running external CAS, from the path ```/etc/cas/saml/idp-metadata.xml```
+        - Assertions : false
+        - Signed request: false.
+      - On vitamui, after creating the provider (SSO list) : 
+        - we download the ```SPS-metadata.xml``` file and copy it in the vm of external CAS, in any directory, 
+            example ```/vitam/some-path-of-cas/SPS-metadata.xml```.
```
Comment (ghost) [inline cas/cas-server/README.md]: remove /vitam

File: cas/cas-server/README.md [hunk 13]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+            example ```/vitam/some-path-of-cas/SPS-metadata.xml```. 
+        - create a new resource file on the project external CAS , example: ```saml-{{my-env-ui}}.json``` in the directory:
+             src/main/resources/services with the following content:
+        ```json
+           {
+          "@class" : "org.apereo.cas.support.saml.services.SamlRegisteredService",
+          "name" : "SAMLService",
+          "id" : 1,
+          "evaluationOrder" : 1,
+          "metadataLocation" : "/vitam/some-path-of-cas/SPS-metadata.xml",
```
Comment (ghost) [inline cas/cas-server/README.md]: idem

File: cas/cas-server/README.md [hunk 14]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+          "skipGeneratingTransientNameId": true,
+          "serviceId" : "https://{{my-env-ui}}.env.programmevitam.fr/cas/login/{{technical-provider-id}}"
```
Comment (ghost) [inline cas/cas-server/README.md]: https://vitamui_host/cas/login

File: cas/cas-server/README.md [hunk 15]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+          }
+        ``` 
+        We have to check the following important informations: 
+        **my-env-ui** is the name of test environment ex: int.
```
Comment (ghost) [inline cas/cas-server/README.md]: fix

File: cas/cas-server/README.md [hunk 16]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+        **some-path-of-cas** is a directory path on the CAS external path vm.
+        **technical-provider-id** is the ```technicalName``` of the provider in providers collection in vitamui db created before.
+    - run ```mvn clean package``` on the project.
+    - run ```java -jar target/cas.war```
+
+- You need to restart the CAS service on the test environment (ex: INT), after adding the provider.
```
Comment (ghost) [inline cas/cas-server/README.md]: int*

File: cas/cas-server/README.md [hunk 17]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+- Create a user into the organisation with email address ```myusernam@mydomainmail.fr``` 
+- We can test the SAML delegated authentication using the external CAS. 
+      
+
+## Auto provisionning:
+
+The auto provisioning is a feature enabling either the updating or creation of users following authentication from an external IdP.
```
Comment (ghost) [inline cas/cas-server/README.md]: The auto provisioning allows??! the creation or update of users following authentication from an external IdP.

File: cas/cas-server/README.md [hunk 18]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+It calls an API to develop to provision a user based on their primary email address, which is the main authentication information.
```
Comment (ghost) [inline cas/cas-server/README.md]: When enabled, the auto-provisioning of user calls an ad-hoc back-end API that returns user information based on their primary email address.

File: cas/cas-server/README.md [hunk 19]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+      client:
+        secure: false
+        ssl-configuration:
+          truststore:
+            key-path: /vitamui/conf/iam-internal/truststore_server.jks
+            key-password: AJ2Ft14CQHiU3eegIAlPqxPRp5uLNMizGadu8SficFja7nQN
+          hostname-verification: false
+```
+- The value of the parameter ```idp-identifier``` should be the ```_id``` of the provider from the collection iam->providers
+  - The parameter ```uri``` is the url of the provisioning api that should return a user when it's called with request parameter ```email```
```
Comment (ghost) [inline cas/cas-server/README.md]: should return user information

File: cas/cas-server/README.md [hunk 20]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+        "zipCode": "Some ZIP code",
+        "city": "Some town",
+        "country": "Some country"
+        },
+        "siteCode": "Some site",
+        "internalCode": "Some internal code"
+        }
+      ```
+    - ***Important*** 
+      - The information ```unit``` is very important for provisoning fetaure, this value is the main information to
```
Comment (ghost) [inline cas/cas-server/README.md]: fetaure --> feature

File: cas/cas-server/README.md [hunk 21]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+             {
+                _id: '662fd94ed0092f16fcadf37507657badae834cfda87af1110d2de0f2a821a60f',
+                identifier: '134',
+                name: 'Groupe provisionning',
+                enabled: true,
+                profileIds: [***],
+                units: [
+                'VITAM'
+                ],....
+            }
```
Comment (ghost) [inline cas/cas-server/README.md:436]: fix json please

File: tools/docker/external-idp-cas/readme.md [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+    preferredJwsAlgorithm: 'ES256',
+    useState: true,
+    useNonce: true,
+    usePkce: false,
+    protocoleType: 'OIDC',
+    customerId: '5c7927af7884583d1ebb6e7a74547a15e35d431599d976a9708eb12d6c5e56c9',
+    _class: 'providers'
+}
+
+```
```
Comment (ghost) [inline tools/docker/external-idp-cas/readme.md:54]: link to cas readme

File: tools/docker/external-idp-cas/run-dev.sh [hunk 23]
```diff
@@ -1,52 +1,90 @@
...
-# Check if access_token is not empty
-if [ -n "admin_access_token" ]; then
+# Function to obtain access token
+get_access_token() {
+    response=$(curl --location "$KEYCLOAK_API_URL/realms/$REALM/protocol/openid-connect/token" \
+               --header 'Content-Type: application/x-www-form-urlencoded' \
+               --data-urlencode "grant_type=client_credentials" \
+               --data-urlencode "client_id=$ADMIN_CLIENT_ID" \
+               --data-urlencode "client_secret=$ADMIN_CLIENT_SECRET")
+    echo "$response" | jq -r '.access_token'
```
Comment (ghost) [inline tools/docker/external-idp-cas/run-dev.sh:77]: requires jq?

File: cas/cas-server/README.md [hunk 24]
```diff
@@ -147,14 +147,280 @@ https://dev.vitamui.com:8080/cas/login).
...
+- Use State: Default to `true`
+- Use Nonce: Default to `true`
+- Use PKCE: Default to `false`
+- Custom Settings: These are customized parameters of the OIDC client if needed.
+
+**Warnings:**
+
+To ensure these configurations work properly, there are certain rules to follow:
+
+- you need to add the CAS Url and Vitamui Url to OIDC Client into the fields:
```
Comment (ghost) [inline cas/cas-server/README.md:175]: Examples to give + cleanup env.progra...

**Review by ghost — APPROVED** [review]

**Review by marob — APPROVED** [review]

**Pull request merged**